subprojects/model-core/src/main/java/org/gradle/model/internal/core/NodeBackedModelSet.java

```diff
@@ -71,17 +71,22 @@ public String toString() {
     }
 
     @Override
-    public void create(final Action<? super T> action) {
+    public void create(Action<? super T> action) {
+        create(getElementTypeClass(), action);
+    }
+
+    @Override
+    public <S extends T> void create(Class<S> type, final Action<? super T> action) {
         state.assertCanMutate();
 
-        String name = String.valueOf(modelNode.getLinkCount(ModelNodes.withType(elementType)));
+        String name = String.valueOf(modelNode.getLinkCount(ModelNodes.withType(type)));
```

Comment:
See the [Model paths](https://docs.gradle.org/current/userguide/software_model.html#N179BB) section of the user guide for a quick but accurate explanation.

In this particular case, we generate `ModelSet` elements paths by appending their index. So a `ModelSet` at `things` path will have elements at `things.0`, `things.1` etc.. That's why we need to count the elements with the base element type.